onnxruntime/core/providers/cpu/tensor/expand.cc

```diff
@@ -0,0 +1,179 @@
+         output_shape_iter != output_shape.rend()) {
+    if (*input_shape_iter != *output_shape_iter) {
+      if (1 == *output_shape_iter) {
+        *output_shape_iter = *input_shape_iter;
+      } else if (1 != *input_shape_iter) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "invalid expand shape");
+      }
+    }
+    input_shape_iter++;
+    output_shape_iter++;
+  }
+
+  TensorShape output_tensor_shape(output_shape);
+  auto* output_tensor = context->Output(0, output_tensor_shape);
+  auto* output_data = output_tensor->template MutableData<T>();
+  auto* output_dims = output_shape.data();
+  auto output_dims_size = static_cast<int64_t>(output_shape.size());
+  auto max_dims_size = std::max(input_dims_size, output_dims_size);
+  std::unique_ptr<int64_t[]> input_dim_group{new int64_t[max_dims_size]};
+  std::unique_ptr<int64_t[]> output_dim_group{new int64_t[max_dims_size]};
+  std::unique_ptr<int64_t[]> expand_dim_size{new int64_t[max_dims_size]};
+  auto dim_group_start = max_dims_size;
+
+  for (int64_t input_dims_iter = input_dims_size - 1,
+               output_dims_iter = output_dims_size - 1,
+               last_dim_size = 1,
+               input_count = 1,
+               output_count = 1;
+       output_dims_iter > -1;
+       input_dims_iter--, output_dims_iter--) {
+    auto input_dim = input_dims_iter > -1 ? input_dims[input_dims_iter] : 1;
+    auto output_dim = output_dims[output_dims_iter];
+
+    input_count *= input_dim;
+    output_count *= output_dim;
+
+    if (0 == input_count || 0 == output_count) {
+      return Status::OK();
+    }
+
+    if (input_dim == 1 && output_dim > 1 || output_dims_iter == 0) {
+      --dim_group_start;
+      input_dim_group[dim_group_start] = input_count;
+      output_dim_group[dim_group_start] = output_count;
+      expand_dim_size[dim_group_start] = output_count / input_count / last_dim_size;
+      last_dim_size *= expand_dim_size[dim_group_start];
+    }
+  }
+
+  auto distribute_count = input_dim_group[dim_group_start] / input_dim_group[max_dims_size - 1];
+  std::vector<int64_t> output_offsets(distribute_count, 0);
+  int64_t copy_len = input_dim_group[max_dims_size - 1];
+  auto copy_byte = copy_len * sizeof(T);
+
+  auto distribute_fn =
+    [&](ptrdiff_t i) {
+      auto input_offset = i * copy_len;
+      int64_t output_offset = 0;
+      for (auto j = dim_group_start + 1, remains = input_offset; j < max_dims_size; ++j) {
+        auto current_count = remains / input_dim_group[j];
+        output_offset += current_count * output_dim_group[j];
+        remains = remains % input_dim_group[j];
+      }  //for
+      memcpy(output_data + output_offset, input_data + input_offset, copy_byte);
+      output_offsets[i] = output_offset;
+    };  //distribute_fn
+
+  auto per_thread_tasks =
+    distribute_count / concurrency::ThreadPool::DegreeOfParallelism(context->GetOperatorThreadPool());
+
+  if (per_thread_tasks > 4) {
```

Comment:
A small note indicating how you arrived at this threshold would help future maintainers.

Reply:
(distribute_count,1,128) - > (distribute_count,10,128)
distribute_count | time(us) 
---- | ---- 
8 | 48
16 | 32
32 | 52
64 | 82
128 | 124
256 | 459
512 | 634